src/pages/create/Opportunity.tsx

```diff
@@ -0,0 +1,162 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import {
+  Button,
+  H0,
+  H2,
+  Stack,
+  useDeskproAppClient,
+  useInitialisedDeskproAppClient,
+} from "@deskpro/app-sdk";
+import { zodResolver } from "@hookform/resolvers/zod";
+import { useEffect, useState } from "react";
+import { useForm } from "react-hook-form";
+import { useNavigate, useParams } from "react-router-dom";
+import { z, ZodTypeAny } from "zod";
+
+import { getObjectMeta, postData } from "../../api/api";
+import { useQueryWithClient } from "../../hooks";
+import { QueryKey } from "../../query";
+import { getMetadataBasedSchema } from "../../schemas/default";
+import opportunityJson from "../../resources/default_layout/opportunity.json";
+import { FieldMappingInput } from "../../components/FieldMappingInput/FieldMappingInput";
+import { Field } from "../../api/types";
+
+const nonUsableFields = ["AccountId", "CreatedDate", "CreatedById"];
+
+const requiredFields = ["Name", "StageName", "CloseDate"];
+
+export const CreateOpportunity = () => {
+  const navigate = useNavigate();
+
+  const [submitting, setSubmitting] = useState(false);
+  const [schema, setSchema] = useState<ZodTypeAny>(z.object({}));
+  const { object, parentId } = useParams();
+
+  const { client } = useDeskproAppClient();
+  const {
+    register,
+    formState: { errors },
+    handleSubmit,
+    setValue,
+    watch,
+  } = useForm<any>({
+    resolver: zodResolver(schema),
+  });
+  useInitialisedDeskproAppClient((client) => {
+    client.deregisterElement("salesforcePlusButton");
+    client.setTitle("Add Opportunity");
+  });
+
+  useEffect(() => {
+    if (parentId) {
+      setValue(object as string, parentId);
+    }
+  }, [parentId, setValue, object]);
+
+  useEffect(() => {
+    register("StageName", {
+      required: true,
+    });
+    register("CloseDate", {
+      required: true,
+    });
+  });
+
+  const opportunityMetadata = useQueryWithClient(
+    [QueryKey.OPPORTUNITY_METADATA, "Opportunity"],
+    (client) => getObjectMeta(client, "opportunity")
+  );
+
+  useEffect(() => {
+    if (!opportunityMetadata.data) return;
+
+    setSchema(
+      getMetadataBasedSchema(opportunityMetadata.data.fields, {
+        Name: z.string().min(1),
```

Comment:
We always need to build the validation schema from sObject metadata as we can't guarantee these fields will exist or be mandatory.

There shouldn't be any circumstance where there are hard coded sObject fields.